docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+
+**OPEN QUESTION:** Should we just make the registration automatic, as part of
+calling `.GenerateSource`? Its trickier to plumb through behind the scenes, but
+I don't see any obvious use case for an author to create a node and not register
+it? I've footgunned myself already with this where I forgot to register the
+output and wondered why nothing was happening.
+
+**OPEN QUESTION** Should we have `GenerateBatch...` versions of the output
+methods? This can already be achieved by the author calling `BatchTransform`
+before calling `Generate...` so would just be a helper, but seems common enough
+that it could be useful.
+
+### Simple example
+
+Putting together the various steps outlined above, an example incremental
+generator might look like the following:
+
+```csharp
+[Generator(LanguageNames.CSharp)]
+public class MyGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext initContext)
+    {
+        initContext.RegisterExecutionPipeline((context) =>
+        {
+            // get the additional text source
+            var additionalTexts = context.Sources.AdditionalTexts;
+
+            // apply a 1-to-1 transform on each text, extracting the path
+            var transformed = additionalTexts.Transform((text) => text.Path);
+
+            // batch the collected file paths into a single collection
+            var batched = transformed.BatchTransform(texts => texts);
+
+            // create an output that takes the file paths from the above batch and makes some user visible syntax
+            var output = batched.GenerateSource((sourceContext, filePaths) =>
+            {
+                sourceContext.AddSource("additionalFiles.cs", @"
+namespace Generated
+{
+    public class AdditionalTextList
+    {
+        public static void PrintTexts()
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}
+");
+            });
+
+            // register the output context.RegisterOutput(output); }); } }
+```
+
+## Advanced Implementation
+
+### Combining and filtering
+
+While the transformation steps outlined above allow a user to create simple
+generators from a single source of data, in reality it is expected that an
+author will need a way to take multiple sources of data and combine them.
+
+There exists an extension method `Combine` that allows the author to take one
+data source and merge it with another, for example, extracting a set of types
+from the compilation and using them to generate something on a per additional
+file basis.
+
+```csharp
+public static partial class IncrementalValueSourceExtensions
+{
+    // join 1 => many ((source1[0], source2), (source1[0], source2), ...)
+    public static IncrementalValueSource<(T, IEnumerable<U>)> Combine<T, U>(this IncrementalValueSource<T> source1, IncrementalValueSource<U> source2) => ...
```

Comment:
i'm on the fence if we should have names for this tuple.  maybe like `source1Value, and source2Values`... something to indicate that they're not singular.